### ISSUE TITLE
Announcement: I'm making a PyWinAutoUI (graphical) GUI with advanced features

### ISSUE THREAD

**Issue by enjoysmath** [pywinauto/pywinauto#906]
My end goal is to do on-screen lessons with overlayed arrows as done on bubble.is (and their set of lessons), but since I don't want to always be coding to create a simple enough lesson, I thought I should give PyWinAuto a GUI frontend that detects / records clicks in .py script format.

Here's a screenshot:
![image](https://user-images.githubusercontent.com/1606391/78190404-cb818b80-7428-11ea-9db9-e843b72b09ff.png)

I'm assuming I should post back here when it's ready.

**Comment by vasily-v-ryabov** [comment]
Hi @enjoysmath nice to see efforts to extend pywinauto ecosystem. We're open to all new ideas and initiatives.

Maybe it's worth telling you about our own efforts:
 * [py_inspect](https://github.com/pywinauto/py_inspect) is a simple GUI inspect tool which is only 150 lines of code using PyQt5. We hope to extend it with advanced recording features in the future, but at the moment there are no people who is ready to do this job.
 * [Automatic script generator](https://github.com/pywinauto/pywinauto/pull/701) is a command line tool that uses all possible hooks and event subscriptions to record your actions into a pywinauto script with text properties where possible. This is not a stupid coordinate recorder/player, but a generator of reliable and maintainable scripts. Unfortunately this 2-years work is paused, because we want to finish Linux and macOS support as well as some re-factoring to make the whole library interface more simple and consistent.

The second tool could be integrated into py_inspect GUI with some additions: single step recording, Undo/Redo and other recorded script editor features. It would be nice to discuss all these things with community.

**Comment by enjoysmath** [comment]
@vasily-v-ryabov  I have already integrated PyInspect efforts into my code.  Unfortunately, I'm new at github and don't know how to recover an old repository (because currently `pywinauto/recorder.py` is not present).  I have already started on a recorder as well, but it would be nice to use an official one coded by someone who isn't a complete newbie at pywinauto :)

**Comment by enjoysmath** [comment]
Here's another screenshot showing progress:

![image](https://user-images.githubusercontent.com/1606391/78318323-4f1aa580-7519-11ea-8069-8f514c55b4c4.png)

I was able to use the multiprocessing library and run the recording on a separate process.  So sockets are used to communicate back to the GUI process.  It's running without blocking the GUI and doesn't seem to crash anything now.  Now just have to convert the events into code statements.

I didn't use the `sandbox/recorder.py` completely since I was using `pywinauto.win32_hooks` library while the recorder is old and uses pyHook which has to be replaced by pyWinhook.  I might have an option to select which one to use.  I will use the existing recorder code to inspire the translation to code calls of each of the events.

**Comment by vasily-v-ryabov** [comment]
No, sandbox repository is very very old! Please don't use it.
I've already posted the link to pull request which contains instructions and code difference:
https://github.com/pywinauto/pywinauto/pull/701

**Comment by vasily-v-ryabov** [comment]
Regarding selling the binaries... of course, you're free to use pywinauto core library for commercial purpose since all the licenses permit it. No problem about that. We can't sponsor this work as there are no donations from community last year. So I completely understand your wish.

The only thing I take care of is a license of your source code. If it's GPL, we can't integrate it into pywinauto core library or into py_inspect. So we will have to re-implement it on our side. I prefer BSD 3-clause or MIT license, but in some cases LGPL might be OK.

If you were so glad to provide pull requests to py_inspect repository, that would be really cool.

**Comment by vasily-v-ryabov** [comment]
Hope it's easy to clone @cetygamer 's fork repository and his branch with the recorder to try it out (just click at the source branch in the pull request #701). Feel free to ask me about the architecture of the recorder.

**Comment by vasily-v-ryabov** [comment]
Sorry for delayed replies. I have a good main job and we work remotely for this quarantine period. So I have no commercial interest for this project in nearest years.

I will monitor this issue in a high priority order and can promise to reply once a day. So that you can spend your time in the most efficient direction.

And thanks for all your efforts!

**Comment by enjoysmath** [comment]
@vasily-v-ryabov  Thank you for all of your prompt responses!  I'm not sure how to do a pull request for py_inspect.  Also I totally broke up `py_inspect`'s code into modules and it runs on a thread as well.

According to this FAQ: https://www.riverbankcomputing.com/commercial/license-faq 
I am limited by PyQt5 library to GPL.

According to this: https://www.riverbankcomputing.com/commercial/buy
PyQt5 is only $550 which is a great deal less than the thousand's I was imagining.  So I might end up going closed source.

**Comment by enjoysmath** [comment]
@vasily-v-ryabov I need more clarification for "clicking at the source branch" on that #701 page.

**Comment by enjoysmath** [comment]
@vasily-v-ryabov It would be crazy to try to get that old code to work.  I cannot seem to bring up that branch in TortoiseGit.  So unless you can provide me with a zip file of or method of checking out code that would plug into the most recent version of pywinauto, that approach is ruled out.   

I am happy with a simpler approach that I can code from scratch (using latest release of pywinauto) and that will meet my needs.  My end goal is a set of on-screen lessons using the apps Qt Creator, WingWare, and Chrome browser, and PyQt5.  By this I mean the user is directed where to click, where to type etc (like [bubble.is lessons](https://bubble.io/lessons)). The lessons will be created using a "Create Lesson" feature of PyWinAutoUI.

**Comment by vasily-v-ryabov** [comment]
You can fork [cetygamer/pywinauto](https://github.com/cetygamer/pywinauto) repository to your GitHub account by pressing "Fork" and waiting for a few minutes.

Then you can clone the repo like this (just replace "cetygamer" with "enjoysmath") in cmd.exe or in GitBash:
```
git clone --branch dev-script-generator https://github.com/cetygamer/pywinauto.git .\pywinauto_fork
```
After the repo is cloned let's change working directory to it and try recorder:
```
cd .\pywinauto_fork
python.exe .\pywinauto\recorder\recorder_cli.py -b uia --cmd .\apps\WPF_samples\WpfApplication1.exe -v -o recorded_wpf.py
```

Some steps might be not recorded with tracebacks in console because it is still prototype. For example, combo box actions is one of the complicated actions that were not finished. But some simple actions are recorded. Actions on side windows (not in the target app) are not recorded.

**Comment by enjoysmath** [comment]
@vasily-v-ryabov  Thank you for helping with git and what to git.  I've got it running and it's throwing some errors, probably because I'm using Python 3.8.    I will get it through the errors.

Question: why do they target a single application when doing the recording?

I was able to do `pywinauto.backend.registry.backends["uia"].element_info_class()` or something similar to get the "root of all things" for which I was also able to record clicks.  If targeting specific apps is best, I can code it to target a small set of apps.

**Comment by enjoysmath** [comment]
@vasily-v-ryabov should I work from cetygamer's or pywinauto's repo to get the recorder up-and-running again?

**Comment by enjoysmath** [comment]
Another question:  I'm changing `..` to pywinauto in a lot of places, usually this is because the way I'm running the source file is not correct.  So what is the way in which one normally runs the recorder say from a pywinauto user's perspective?  That is what I need since these import changes will likely end up breaking things when I switch to the standard way.

**Comment by enjoysmath** [comment]
@vasily-v-ryabov I'm not sure what to do next as far as getting this to work with the latest pywinauto.  I'll assume I just fork pywinauto/pywinauto and begin manually migrating the recorder code over from cetygamer/pywinauto fork.  I think this is a first step, as it would be weird to depend on both repositories and there would be conflicts involving the generated code.

**Comment by enjoysmath** [comment]
@vasily-v-ryabov 
Any idea how to fix this error:

![image](https://user-images.githubusercontent.com/1606391/78411852-0a991380-75c6-11ea-8bbe-c80280d316ea.png)

I made minimal changes after copying over the `recorder` folder to other parts of the code in order to get it to run.  See master branch of:  https://github.com/enjoysmath/pywinauto

**Comment by airelil** [comment]
>I fixed that line by using GetModuleHandleW instead of GetModuleHandleA but I wrapped the original in a try/except block so it should work as original code except if in my case it breaks then it will use the unicode version (the 'W' version) of GetModuleHandle func. 

Was fixed in another branch: https://github.com/pywinauto/pywinauto/commit/201bd217f603e6902ada920558fe613860cbf301

**Comment by enjoysmath** [comment]
> > I fixed that line by using GetModuleHandleW instead of GetModuleHandleA but I wrapped the original in a try/except block so it should work as original code except if in my case it breaks then it will use the unicode version (the 'W' version) of GetModuleHandle func.
> 
> Was fixed in another branch: [201bd21](https://github.com/pywinauto/pywinauto/commit/201bd217f603e6902ada920558fe613860cbf301)

How did you find this out?

**Comment by enjoysmath** [comment]
@airelil I forked and cloned pywinauto/pywinauto so if it was fixed and atspi is the default branch, then why didn't my checked out code show that updated code?  So, how would I bring in the changes to my code using git checkout?

**Comment by airelil** [comment]
As I understand, you worked on a "recorder" branch that probably is quite behind the rest of the code. This is a complicated feature and the work on it has been done in parallel by other contributors. Usually, once the feature is stable enough it's merged into the main branch. Currently we have master as our stable branch and atspi as our development branch where we consolidate all upcoming features for the next release. To align with atspi branch you need to run "git merge", but at this stage with such a big difference between two branches (many files have been moved under different locations) it will require to resolve many conflicts in different areas of the code.

**Comment by airelil** [comment]
BTW, @enjoysmath, did you manage to run unit tests?

**Comment by vasily-v-ryabov** [comment]
@enjoysmath sorry I was unable to be near PC during the weekend. I'm going to merge `atspi` branch into `dev-script-recorder`. This is pretty complicated merge because current state of recorder is based on master branch while `atspi` has some deep changes with moving the code to OS specific modules (master branch is Windows only). Also I will push the recorder branch to the upstream (pywinauto/pywinauto). Hope it will make your life simpler.

**Comment by vasily-v-ryabov** [comment]
I've almost finished the merge in my fork. Will push it to upstream when the tests are passed.

At the same time it is useful to learn how to add remote repositories to the local clone and sync from another fork or from upstream:
https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork

**Comment by vasily-v-ryabov** [comment]
Well, the updated "atspi-based" branch in my fork looks in the same state as "master-based" branch:
 * [AppVeyor tests for master-based branch](https://ci.appveyor.com/project/pywinauto/pywinauto/builds/25149402)
 * [AppVeyor tests for atspi-based branch](https://ci.appveyor.com/project/pywinauto/pywinauto-8b7t7/builds/31976092)

I've done for today. The tests are still having few failures. But I've pushed the branch to upstream as well. Can't promise to fix them quickly. But it's possible to send pull requests to this branch at least. ;)

**Comment by vasily-v-ryabov** [comment]
Regarding several applications monitoring... of course, it's possible. We didn't think in details about the whole mechanism, but we plan to implement spawning child processes detection in `Application` object in the core library. We may think to add a list of processes or handle several Application objects at the same time. This is interesting point to discuss.

**Comment by vasily-v-ryabov** [comment]
Not sure about dependencies between py_inspect and pywinauto. I thought it could be integrated so: py_inspect uses pywinauto as a package installed into Python. But temporary prototype could be implemented in the way you like. The command line recorder is planned to be a part of pywinauto repository to avoid maintaining versions compatibility between recorder and core library.

**Comment by jjbright** [comment]
This looks interesting to me, can I know how to use the code generator for pywinauto  and how to run the pywinauto GUI @enjoysmath @vasily-v-ryabov @airelil 
I'm struggling a lot using the print_control_identifiers() and find the path for the controls in the screen.

The control types are UIA

**Comment by enjoysmath** [comment]
@jjbright Follow the instructions [here](https://github.com/pywinauto/pywinauto/issues/906#issuecomment-608535930) (above):

If you want to run the debugger from the commandline you have to specify the full absolute path of the executable app.

**Comment by jjbright** [comment]
> @jjbright Follow the instructions [here](https://github.com/pywinauto/pywinauto/issues/906#issuecomment-608535930) (above):
> 
> If you want to run the debugger from the commandline you have to specify the full absolute path of the executable app.

I tried like below 
`
C:\Python\Python37>python .\pywinauto\recorder\recorder_cli.py -b uia --cmd C:\Windows\System32\notepad.exe -v -o recorded_wpf.py
python: can't open file '.\pywinauto\recorder\recorder_cli.py': [Errno 2] No such file or directory

C:\Python\Python37>
`

**Comment by enjoysmath** [comment]
@jjbright Did you git the source code of that other repository?  The recorder code is not part of this repo afaik.

**Comment by jjbright** [comment]
> @jjbright Did you git the source code of that other repository? The recorder code is not part of this repo afaik. (See instructions again)

@enjoysmath I followed the same steps as mentioned, so now I get **ModuleNotFoundError**

`
C:\Users\jjbright\source\repos\pywinauto_fork>python.exe .\pywinauto\recorder\recorder_cli.py -b uia --cmd C:\Windows\System32\notepad.exe -v -o recorded_wpf.py
Traceback (most recent call last):
  File ".\pywinauto\recorder\recorder_cli.py", line 17, in <module>
    from pywinauto.recorder.uia.uia_recorder import UiaRecorder
ModuleNotFoundError: No module named 'pywinauto.recorder'
`
